### PR TITLE
Fix detection of unmappable root domains

### LINKED ISSUE #1950: CA root domain mapping result depends on fusion output order
### 🐛 Describe the bug

`ComputeAtRootDomainMap` gives different mapping results in the below 2 fusions, and note that the only difference is the output ordering.

@csarofeen @naoyam @zasdfgbnm 
```cpp
TEST_F(NVFuserTest, FusionRootMapReproPart1_CUDA) {
    Fusion fusion;
    FusionGuard fg(&fusion);
    auto t21 = makeSymbolicTensor(3);
    auto t5 = makeSymbolicTensor(3);
    auto t1 = makeSymbolicTensor(3);
    
    fusion.addInput(t21);
    fusion.addInput(t5);
    fusion.addInput(t1);

    auto t26 = set(t21);
    auto t29 = mul(t5, t26);
    auto t19 = mul(t5, t1);
    auto t28 = mul(t19, t26);
    auto t35 = sum(t28, {2});
    auto t38 = broadcast(t35, {false,false, true});
    auto t41 = mul(t26, t38);
    
    // Only output order is different in the 2 fusion defs
    fusion.addOutput(t41);
    fusion.addOutput(t19);
    fusion.addOutput(t29);
    
    ComputeAtRootDomainMap root_map;
    root_map.build();

        // vvvvv The check polarity is different
    TORCH_CHECK(root_map.canMap(t29->domain(), t29->axis(-1), t41->domain(), t41->axis(-1)));
}

TEST_F(NVFuserTest, FusionRootMapReproPart2_CUDA) {
    Fusion fusion;
    FusionGuard fg(&fusion);
    auto t21 = makeSymbolicTensor(3);
    auto t5 = makeSymbolicTensor(3);
    auto t1 = makeSymbolicTensor(3);
    
    fusion.addInput(t21);
    fusion.addInput(t5);
    fusion.addInput(t1);

    auto t26 = set(t21);
    auto t29 = mul(t5, t26);
    auto t19 = mul(t5, t1);
    auto t28 = mul(t19, t26);
    auto t35 = sum(t28, {2});
    auto t38 = broadcast(t35, {false,false, true});
    auto t41 = mul(t26, t38);

        // Only output order is different in the 2 fusion defs
    fusion.addOutput(t29);
    fusion.addOutput(t41);
    fusion.addOutput(t19);
    
    ComputeAtRootDomainMap root_map;
    root_map.build();

        // vvvvv The check polarity is different
    TORCH_CHECK(!root_map.canMap(t29->domain(), t29->axis(-1), t41->domain(), t41->axis(-1)));
}
```

Looks like it is the unmappable dimension logic that was traversal order dependent. Could possibly use a two-traversal approach, just a thought...

### Versions

devel branch

### PR DESCRIPTION
**Pull request by naoyam**
`ComputeAtRootDomainMap` flags domains that should not be mapped due to
reductions. Previously, checking if a domain potentially causes an
invalid mapping is only done with one domain in each group of domains
that are found to be mappable so far. That's not actually sufficient as
the unmappable domain set is created just once with no root mapping
information. The fix is to check all consumer domains of a producer
tensor. A small other fix is also done to address a different problem
discovered after the first fix.

Fixes #1950

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 1]
```diff
@@ -339,30 +343,41 @@ void UnmappableReductionDomains::handle(WelfordOp* op) {
...
+
+    // A consumer domain that is an input to the reduction domain
+    const DomainKey& input_to_reduction = *it;
+
+    // Check if mapping input_to_reduction with the other domains in
+    // consumer_domains. If there's a domain that is a consumer of the
+    // reduction, they must not be mapped together
     for (const auto& consumer_domain : consumer_domains) {
-      if (consumer_domain == consumer_domain_with_reduction) {
+      if (consumer_domain == input_to_reduction) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:380]: Just additional comments and cleanup

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 2]
```diff
@@ -1070,26 +1107,14 @@ bool ComputeAtRootDomainMapBuilder::safeToMap(const DomainKeySet& domains) {
...
-              return root_map_.canMap(domain, unique_dom);
-            })) {
-      unique_domains.push_back(domain);
-    }
-  }
+
   // Can't map if reduction output domains would be mapped
-  if (incompatible_domains_.isReductionOutputMapped(
-          unique_domains, root_map_) &&
+  if (incompatible_domains_.isReductionOutputMapped(domains, root_map_) &&
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:1112]: This is the main fix. Just checking unique domains isn't sufficient.

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 3]
```diff
@@ -297,6 +297,10 @@ void UnmappableReductionDomains::handleReductionOutput(TensorView* out_tv) {
   auto use_chains = DependencyCheck::getAllUseChains(out_tv);
   for (const auto& chain : use_chains) {
     for (const auto& tv : ir_utils::filterByType<TensorView>(chain)) {
+      // Do not include the tensor itself in its consumers
+      if (tv == out_tv) {
+        continue;
+      }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:303]: This is another related fix. Here, we are gathering IterDomains of all consumer tensors of a tensor with a reduction IterDomain. For this logic, we don't want to have the same tensor as its consumer.

**Review by shmsong — APPROVED** [review]
LGTM. Thanks for fixing this!

**Pull request merged**